### PR TITLE
Update ErrorGroupCallback to use IReadOnlyDictionary with API changes

### PR DESCRIPTION
**Pull request by jaffinito**
## Description

Makes some changes to Error Groups based on feedback:
* Update ErrorGroupCallback to use IReadOnlyDictionary with API changes
* Update ErrorData to keep the Exception
* Make GetAllAttributeValuesDic return an IDictionary
* Add the Trucated (cleaned) StackTrace and the Exception to the attribvute
* Only convert the IDictionary to IReadOnlyDictionary after adding the stack trace and Exception, if present
* If there is no callback do not get the attributes
* Add supportability metric with timing of SetErrorGroup in ErrorEventMaker and ErrorTraceMaker called *SetErrorGroup

# Author Checklist
- [x] Unit tests, Integration tests, and Unbounded tests completed
- [x] Performance testing completed with satisfactory results (if required)
- [ ] [Agent Changelog](/src/Agent/CHANGELOG.md) or [Lambda Agent changelog](/src/AwsLambda/CHANGELOG.md) updated 

# Reviewer Checklist
- [ ] Perform code review
- [ ] Pull request was adequately tested (new/existing tests, performance tests)
- [ ] Review Changelog

### CODE REVIEW

File: src/Agent/NewRelic/Agent/Core/Attributes/AttributeValueCollection.cs [hunk 1]
```diff
@@ -138,6 +148,26 @@ public IEnumerable<IAttributeValue> GetAttributeValues(AttributeClassification c
...
         }
 
+        public IDictionary<string, object> GetAllAttributeValuesDic()
+        {
+            var result = new Dictionary<string, object>();
+            foreach (var classification in _allClassifications)
+            {
+                foreach (var attribVal in GetAttribValuesImpl(classification))
+                {
+                    if (classification == AttributeClassification.UserAttributes && _reservedProperties.Contains(attribVal.AttributeDefinition.Name) )
```
Comment (chynesNR) [inline src/Agent/NewRelic/Agent/Core/Attributes/AttributeValueCollection.cs:158]: Rather than having explicit reserved properties, maybe it would be cleaner to loop through the classifications in increasing priority order? user -> intrinsic -> agent?

File: src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs [hunk 2]
```diff
@@ -43,7 +55,40 @@ public ErrorEventWireModel GetErrorEvent(ImmutableTransaction immutableTransacti
...
+
+            using (_agentTimerService.StartNew(SetErrorGroupSupportabilityName))
+            {
+                var callbackAttributes = attribValues.GetAllAttributeValuesDic();
+                if (errorData?.RawException != null)
+                {
+                    callbackAttributes.Add("exception", errorData.RawException);
+                }
+
+                callbackAttributes.Add("stack_trace", GetFormattedStackTrace(errorData));
```
Comment (chynesNR) [inline src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs:88]: Maybe define a constant for the extra attributes being added?

File: src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs [hunk 3]
```diff
@@ -43,7 +55,40 @@ public ErrorEventWireModel GetErrorEvent(ImmutableTransaction immutableTransacti
...
+            {
+                return;
+            }
+
+            using (_agentTimerService.StartNew(SetErrorGroupSupportabilityName))
+            {
+                var callbackAttributes = attribValues.GetAllAttributeValuesDic();
+                if (errorData?.RawException != null)
+                {
+                    callbackAttributes.Add("exception", errorData.RawException);
```
Comment (chynesNR) [inline src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs:85]: Might want to use `TryAdd` just in case there's a user attribute with the same name.

**Comment by chynesNR** [comment]
Looks great!

File: src/Agent/NewRelic/Agent/Core/Attributes/AttributeValueCollection.cs (hunk 1 above)
Comment (nrcventura) [reply src/Agent/NewRelic/Agent/Core/Attributes/AttributeValueCollection.cs:158]: I think that this recommendation is simpler but probably deserves a comment and some unit tests otherwise someone might change the order in the future.

File: src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs (hunk 3 above)
Comment (nrcventura) [reply src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs:85]: Either that or use the `[]=` syntax which also avoids the exception.

File: src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs (hunk 2 above)
Comment (jaffinito) [reply src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs]: Done!

File: src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs (hunk 3 above)
Comment (jaffinito) [reply src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs]: Done!

File: src/Agent/NewRelic/Agent/Core/Attributes/AttributeValueCollection.cs (hunk 1 above)
Comment (jaffinito) [reply src/Agent/NewRelic/Agent/Core/Attributes/AttributeValueCollection.cs]: I updated the existing "_allClassifications" to use the order you mentioned.  I double checked and it everywhere else it was used did not depend on the order.  I also added a few comments with the order and a new test that confirms the order works.

File: src/Agent/NewRelic/Agent/Core/Segments/Segment.cs [hunk 4]
```diff
@@ -135,7 +135,7 @@ public void End()
                 var endTime = _transactionSegmentState.GetRelativeTime();
                 RelativeEndTime = endTime;
 
-                if (Agent.Instance.StackExchangeRedisCache != null && Agent.Instance.StackExchangeRedisCache.Count > 0)
+                if (Agent.Instance?.StackExchangeRedisCache != null && Agent.Instance.StackExchangeRedisCache.Count > 0)
```
Comment (tippmar-nr) [inline src/Agent/NewRelic/Agent/Core/Segments/Segment.cs:138]: This seems unrelated to the rest of this PR. Does it belong here or in a separate PR?

File: src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs [hunk 5]
```diff
@@ -44,13 +57,40 @@ public ErrorEventWireModel GetErrorEvent(ImmutableTransaction immutableTransacti
...
+
         private void SetErrorGroup(ErrorData errorData, IAttributeValueCollection attribValues)
         {
-            _attribDefs.ErrorGroup.TrySetValue(attribValues, errorData.ErrorGroup);
+            if (_configurationService.Configuration.ErrorGroupCallback == null)
+            {
+                return;
+            }
+
+            using (_agentTimerService.StartNew(SetErrorGroupSupportabilityName))
```
Comment (tippmar-nr) [inline src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs:82]: I was thinking we'd put the timer around just the callback invocation. But maybe the little bit of extra work occurring here isn't enough to matter.

File: src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorTraceMaker.cs [hunk 6]
```diff
@@ -103,10 +112,30 @@ private IList<string> GetFormattedStackTrace(ErrorData errorData)
...
         }
 
-        private ErrorTraceWireModel.ErrorTraceAttributesWireModel GetErrorTraceAttributes(ErrorData errorData, IAttributeValueCollection attributes, IList<string> stackTrace)
+        private ErrorTraceWireModel.ErrorTraceAttributesWireModel GetErrorTraceAttributes(IAttributeValueCollection attributes, IList<string> stackTrace)
         {
-            _attribDefs.ErrorGroup.TrySetValue(attributes, errorData.ErrorGroup);
             return new ErrorTraceWireModel.ErrorTraceAttributesWireModel(attributes, stackTrace);
         }
+
+        private void SetErrorGroup(ErrorData errorData,  IList<string> stackTrace,IAttributeValueCollection attribValues)
```
Comment (tippmar-nr) [inline src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorTraceMaker.cs:120]: nit: code formatting, need a space after `stackTrace,`

File: src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs (hunk 5 above)
Comment (chynesNR) [reply src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs:82]: I can see value in both. There could be a potential performance issue around generating the dictionary, but timing just the callback may be a clearer "this problem is definitely on your end" signal.

File: src/Agent/NewRelic/Agent/Core/Segments/Segment.cs (hunk 4 above)
Comment (jaffinito) [reply src/Agent/NewRelic/Agent/Core/Segments/Segment.cs:138]: I might not be required at all.  My unit tests were passing and then out of the blue started to fail at this point with an object null error.  I made the change, restarted VS, and then it was all fine.  I kept it in since it made sense to check both for null, but it is probably safe to undo the change.

File: src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs (hunk 5 above)
Comment (jaffinito) [reply src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs:82]: I waffled back and forth between the two and went with timing most of the work, but if we want to just time the callback, I am totally fine with that.

File: src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorTraceMaker.cs (hunk 6 above)
Comment (jaffinito) [reply src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorTraceMaker.cs:120]: Will fix!

File: src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs (hunk 5 above)
Comment (jaffinito) [reply src/Agent/NewRelic/Agent/Core/Transformers/TransactionTransformer/ErrorEventMaker.cs:94]: Updated to time callback only.

File: src/Agent/NewRelic/Agent/Core/Segments/Segment.cs (hunk 4 above)
Comment (jaffinito) [reply src/Agent/NewRelic/Agent/Core/Segments/Segment.cs]: Removed.

**Review by chynesNR — APPROVED** [review]

**Review by tippmar-nr — APPROVED** [review]
Looks great!
